File: src/main/java/com/stargazerstudios/existence/sonata/entity/Event.java
package com.stargazerstudios.existence.sonata.entity;

import lombok.*;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import javax.persistence.*;
import java.sql.Timestamp;
import java.time.LocalDate;
import java.util.LinkedHashSet;
import java.util.Set;

@Entity
@Table(name = "sonata_events")
@Getter @Setter @NoArgsConstructor
public class Event {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private long id;

    @Column(name = "start_date", columnDefinition = "DATE")
    private LocalDate startDate;

    @Column(name = "end_date", columnDefinition = "DATE")
    private LocalDate endDate;

    @Column(name = "jira_case")
    private String jiraCase;

    @Column(name = "features_on")
    private String featuresOn;

    @Column(name = "features_off")
    private String featuresOff;

    @Column(name = "compiled_sources")
    private String compiledSources;

    @Column(name = "api_used")
    private String apiUsed;

    @Column(name = "created_by")
    private String createdBy;

    @CreationTimestamp
    @Column(name = "creation_date")
    private Timestamp dateCreated;

    @Column(name = "last_changed_by")
    private String lastChangedBy;

    @UpdateTimestamp
    @Column(name = "last_changed_date")
    private Timestamp dateChanged;

    @ManyToOne
    @JoinColumn(name = "system_id")
    private System system;

    @ManyToMany
    @JoinTable(
            name = "events_zones",
            joinColumns = {
                    @JoinColumn(name = "event_id", referencedColumnName = "id")
            },
            inverseJoinColumns = {
                    @JoinColumn(name = "zone_id", referencedColumnName = "id")
            }
    )
    @OrderBy("id")
    private Set<Zone> zones = new LinkedHashSet<>();

    @ManyToMany
    @JoinTable(
            name = "events_event_types",
            joinColumns = {
                    @JoinColumn(name = "event_id", referencedColumnName = "id")
            },
            inverseJoinColumns = {
                    @JoinColumn(name = "event_type_id", referencedColumnName = "id")
            }
    )
    @OrderBy("id")
    private Set<EventType> eventTypes = new LinkedHashSet<>();
}

File: src/main/resources/frontend/existence/src/app/sonata/models/event.ts
export class Event {
  id!: number;
  system!: string;
  zones!: string[];
  zonesStr!: string;
  event_types!: string[];
  event_typesStr!: string;
  start_date!: string;
  end_date!: string;
  jira_case!: string;
  features_on!: string;
  features_off!: string;
  compiled_sources!: string;
  api_used!: string;
  created_by!: string;
  creation_date!: string;
  last_changed_by!: string;
  last_changed_date!: string;

  constructor(object?: any) {
    if (object != undefined) {
      this.id = object.id;
      this.system = object.system;
      this.zones = object.zones;
      this.event_types = object.event_types;
      this.start_date = object.start_date;
      this.end_date = object.end_date;
      this.jira_case = object.jira_case;
      this.features_on = object.features_on;
      this.features_off = object.features_off;
      this.compiled_sources = object.compiled_sources;
      this.api_used = object.api_used;
      this.created_by = object.created_by;
      this.creation_date = object.creation_date;
      this.last_changed_by = object.last_changed_by;
      this.last_changed_date = object.last_changed_date;
    }
  }

  public stringify(): Event {
    if (this.zones.length > 0) {
      for (let i = 0; i < this.zones.length; i++) {
        if (i == 0) {
          this.zonesStr = this.zones[i]
        } else {
          this.zonesStr = this.zonesStr + '-' + this.zones[i];
        }
      }
    }

    if (this.event_types.length > 0) {
      for (let i = 0; i < this.event_types.length; i++) {
        if (i == 0) {
          this.event_typesStr = this.event_types[i]
        } else {
          this.event_typesStr = this.event_typesStr + '-' + this.event_types[i];
        }
      }
    }

    return this;
  }
}
